File: src/main/java/lekkit/scev/network/DisplayPayload.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.network;

import io.netty.buffer.ByteBuf;
import java.util.UUID;
import lekkit.scev.main.ScalarEvolution;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;

/**
 * Server -> Client: framebuffer blit broadcast from a running machine.
 *
 * <p>{@code pixels} is a raw ARGB byte stream ({@code width * height * 4} bytes).
 */
public record DisplayPayload(UUID machineUuid, short width, short height, byte[] pixels)
        implements CustomPacketPayload {

    public static final CustomPacketPayload.Type<DisplayPayload> TYPE =
            new CustomPacketPayload.Type<>(ScalarEvolution.rl("display"));

    public static final StreamCodec<ByteBuf, DisplayPayload> STREAM_CODEC =
            StreamCodec.of(DisplayPayload::encode, DisplayPayload::decode);

    private static void encode(ByteBuf buf, DisplayPayload p) {
        buf.writeLong(p.machineUuid.getMostSignificantBits());
        buf.writeLong(p.machineUuid.getLeastSignificantBits());
        buf.writeShort(p.width);
        buf.writeShort(p.height);
        buf.writeInt(p.pixels.length);
        buf.writeBytes(p.pixels);
    }

    private static DisplayPayload decode(ByteBuf buf) {
        long mb = buf.readLong();
        long lb = buf.readLong();
        short w = buf.readShort();
        short h = buf.readShort();
        int len = buf.readInt();
        byte[] px = new byte[Math.max(0, len)];
        buf.readBytes(px);
        return new DisplayPayload(new UUID(mb, lb), w, h, px);
    }

    @Override public Type<DisplayPayload> type() { return TYPE; }
}

File: src/main/kotlin/lekkit/scev/network/DisplayPayload.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.network

import io.netty.buffer.ByteBuf
import java.util.UUID
import lekkit.scev.core.time.Micros
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload

/**
 * Server → Client: framebuffer blit broadcast from a running machine.
 * `pixels` is the encoded H.264 NAL byte stream for the captured frame
 * (decoded shape: `width * height` BGRA pixels).
 *
 * [ptsMicros] is the capture-time PTS in the per-machine clock shared
 * with [SoundFramePayload]. The client's `MediaClock` keys the jitter
 * buffer by PTS so video presentations stay locked to audio playback.
 *
 * Stream end is signalled by [DisplayDisposePayload], not by an
 * in-band size-0 sentinel — see that class for the rationale.
 */
data class DisplayPayload(
    @get:JvmName("machineUuid") val machineUuid: UUID,
    @get:JvmName("ptsMicros")   val ptsMicros: Micros,
    @get:JvmName("width")       val width: Short,
    @get:JvmName("height")      val height: Short,
    @get:JvmName("pixels")      val pixels: ByteArray,
) : CustomPacketPayload {

    override fun type(): CustomPacketPayload.Type<DisplayPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<DisplayPayload> = payloadType("display")

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, DisplayPayload> = StreamCodec.of(::encode, ::decode)

        /**
         * Java-friendly factory taking a raw `Long` PTS. Kotlin callers
         * use the primary constructor directly with [Micros].
         */
        @JvmStatic
        fun create(
            machineUuid: UUID,
            ptsMicros: Long,
            width: Short,
            height: Short,
            pixels: ByteArray,
        ): DisplayPayload = DisplayPayload(machineUuid, Micros(ptsMicros), width, height, pixels)

        private fun encode(buf: ByteBuf, p: DisplayPayload) {
            buf.writeUuid(p.machineUuid)
            buf.writeLong(p.ptsMicros.value)
            buf.writeShort(p.width.toInt())
            buf.writeShort(p.height.toInt())
            buf.writeSizedBytes(p.pixels)
        }

        private fun decode(buf: ByteBuf): DisplayPayload = DisplayPayload(
            machineUuid = buf.readUuid(),
            ptsMicros   = Micros(buf.readLong()),
            width       = buf.readShort(),
            height      = buf.readShort(),
            pixels      = buf.readSizedBytes(),
        )
    }
}
